# Exploratory Data Analysis

Imports

In [46]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np

### Let's load the data and make a glimpse overview

In [19]:
ratings = pd.read_csv('data/ratings.dat', sep='::', engine='python', names=['UserID', 'MovieID', 'Rating', 'Timestamp'])

In [20]:
ratings.head()

UserID  MovieID  Rating  Timestamp
0       1     1193       5  978300760
1       1      661       3  978302109
2       1      914       3  978301968
3       1     3408       4  978300275
4       1     2355       5  978824291

In [28]:
#let's convert 'Timestamp' column into a more readable view
ratings['Timestamp'] = pd.to_datetime(ratings['Timestamp'], unit='s')

In [29]:
ratings.head()

UserID  MovieID  Rating           Timestamp
0       1     1193       5 2000-12-31 22:12:40
1       1      661       3 2000-12-31 22:35:09
2       1      914       3 2000-12-31 22:32:48
3       1     3408       4 2000-12-31 22:04:35
4       1     2355       5 2001-01-06 23:38:11

In [30]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000209 entries, 0 to 1000208
Data columns (total 4 columns):
 #   Column     Non-Null Count    Dtype         
---  ------     --------------    -----         
 0   UserID     1000209 non-null  int64         
 1   MovieID    1000209 non-null  int64         
 2   Rating     1000209 non-null  int64         
 3   Timestamp  1000209 non-null  datetime64[ns]
dtypes: datetime64[ns](1), int64(3)
memory usage: 30.5 MB


In [82]:
ratings.describe(include='all')

C:\Users\Yurii Shynkarov\AppData\Local\Temp\ipykernel_11544\616227738.py:1: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  ratings.describe(include='all')


UserID       MovieID        Rating            Timestamp
count   1.000209e+06  1.000209e+06  1.000209e+06              1000209
unique           NaN           NaN           NaN               458455
top              NaN           NaN           NaN  2000-11-29 20:06:42
freq             NaN           NaN           NaN                   30
first            NaN           NaN           NaN  2000-04-25 23:05:32
last             NaN           NaN           NaN  2003-02-28 17:49:50
mean    3.024512e+03  1.865540e+03  3.581564e+00                  NaN
std     1.728413e+03  1.096041e+03  1.117102e+00                  NaN
min     1.000000e+00  1.000000e+00  1.000000e+00                  NaN
25%     1.506000e+03  1.030000e+03  3.000000e+00                  NaN
50%     3.070000e+03  1.835000e+03  4.000000e+00                  NaN
75%     4.476000e+03  2.770000e+03  4.000000e+00                  NaN
max     6.040000e+03  3.952000e+03  5.000000e+00                  NaN

In [63]:
ratings.groupby('UserID').agg({'Rating': 'count'}).reset_index().sort_values('Rating', ascending=True)

UserID  Rating
946      947      20
4067    4068      20
2529    2530      20
340      341      20
5257    5258      20
...      ...     ...
1180    1181    1521
1940    1941    1595
4276    4277    1743
1679    1680    1850
4168    4169    2314

[6040 rows x 2 columns]

In [62]:
print("Duplicate entries:", ratings.duplicated().sum())

Duplicate entries: 0


We can confirm that: 

- UserIDs range between 1 and 6040 
- MovieIDs range between 1 and 3952
- Ratings are made on a 5-star scale (whole-star ratings only)
- Each user has at least 20 ratings
- There are no NULL values in all columns
- There are no fully duplicated rows in the table
- The ratings are collected during 2000-11-29 - 2003-02-28 period

In [65]:
users = pd.read_csv('data/users.dat', sep='::', engine='python', names=['UserID', 'Gender', 'Age', 'Occupation', 'Zip-code'])

In [66]:
users.head()

UserID Gender  Age  Occupation Zip-code
0       1      F    1          10    48067
1       2      M   56          16    70072
2       3      M   25          15    55117
3       4      M   45           7    02460
4       5      M   25          20    55455

In [67]:
users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6040 entries, 0 to 6039
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   UserID      6040 non-null   int64 
 1   Gender      6040 non-null   object
 2   Age         6040 non-null   int64 
 3   Occupation  6040 non-null   int64 
 4   Zip-code    6040 non-null   object
dtypes: int64(3), object(2)
memory usage: 236.1+ KB


In [68]:
users.describe(include='all')

UserID Gender          Age   Occupation Zip-code
count   6040.000000   6040  6040.000000  6040.000000     6040
unique          NaN      2          NaN          NaN     3439
top             NaN      M          NaN          NaN    48104
freq            NaN   4331          NaN          NaN       19
mean    3020.500000    NaN    30.639238     8.146854      NaN
std     1743.742145    NaN    12.895962     6.329511      NaN
min        1.000000    NaN     1.000000     0.000000      NaN
25%     1510.750000    NaN    25.000000     3.000000      NaN
50%     3020.500000    NaN    25.000000     7.000000      NaN
75%     4530.250000    NaN    35.000000    14.000000      NaN
max     6040.000000    NaN    56.000000    20.000000      NaN

In [69]:
print('The number of unique users:', users.UserID.nunique())
print('Genders presented in the dataset:', users.Gender.unique())
print('Age groups presented in the dataset:', np.sort(users.Age.unique()))
print('Occupations presented in the dataset:', np.sort(users.Occupation.unique()))

The number of unique users: 6040
Genders presented in the dataset: ['F' 'M']
Age groups presented in the dataset: [ 1 18 25 35 45 50 56]
Occupations presented in the dataset: [ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20]


According to the information from the README.txt file we can create 'occupation_dict' describing each occupation value:

In [70]:
occupation_dict = {
    0: "other",
    1: "academic/educator",
    2: "artist",
    3: "clerical/admin",
    4: "college/grad student",
    5: "customer service",
    6: "doctor/health care",
    7: "executive/managerial",
    8: "farmer",
    9: "homemaker",
    10: "K-12 student",
    11: "lawyer",
    12: "programmer",
    13: "retired",
    14: "sales/marketing",
    15: "scientist",
    16: "self-employed",
    17: "technician/engineer",
    18: "tradesman/craftsman",
    19: "unemployed",
    20: "writer"
}

In [71]:
users['Occupation'] = users['Occupation'].map(occupation_dict)

In [72]:
users.head()

UserID Gender  Age            Occupation Zip-code
0       1      F    1          K-12 student    48067
1       2      M   56         self-employed    70072
2       3      M   25             scientist    55117
3       4      M   45  executive/managerial    02460
4       5      M   25                writer    55455

Let's check if there are new null values occured

In [76]:
users.Occupation.isna().sum()

0

Also, let's check if all values in Age column are valid according to the data description:

In [73]:
print("Users with invalid Age values:", users[~users['Age'].isin([1, 18, 25, 35, 45, 50, 56])])

Users with invalid Age values: Empty DataFrame
Columns: [UserID, Gender, Age, Occupation, Zip-code]
Index: []


In [79]:
print("Duplicate entries in users data:", users.duplicated().sum())

Duplicate entries in users data: 0


We can see that: 
* Gender is denoted by a "M" for male and "F" for female and these are only values in the 'Gender' column
* Age column describes not a precise age but an age group. All the values are valid
* All Occupation values are valid
* There are no NULL values in the dataframe
* There are no duplicates 

In [52]:
movies = pd.read_csv('data/movies.dat', sep='::', engine='python', names=['MovieID', 'Title', 'Genres'], encoding='latin1')

In [53]:
movies.head()

MovieID                               Title                        Genres
0        1                    Toy Story (1995)   Animation|Children's|Comedy
1        2                      Jumanji (1995)  Adventure|Children's|Fantasy
2        3             Grumpier Old Men (1995)                Comedy|Romance
3        4            Waiting to Exhale (1995)                  Comedy|Drama
4        5  Father of the Bride Part II (1995)                        Comedy

In [54]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3883 entries, 0 to 3882
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   MovieID  3883 non-null   int64 
 1   Title    3883 non-null   object
 2   Genres   3883 non-null   object
dtypes: int64(1), object(2)
memory usage: 91.1+ KB


In [59]:
movies.describe(include='all')

MovieID             Title Genres
count   3883.000000              3883   3883
unique          NaN              3883    301
top             NaN  Toy Story (1995)  Drama
freq            NaN                 1    843
mean    1986.049446               NaN    NaN
std     1146.778349               NaN    NaN
min        1.000000               NaN    NaN
25%      982.500000               NaN    NaN
50%     2010.000000               NaN    NaN
75%     2980.500000               NaN    NaN
max     3952.000000               NaN    NaN

Let's check if all genres in the dataset are valid with respect of README.txt dataset description

In [60]:
print("Movies with invalid Genres:")
valid_genres = set(['Action', 'Adventure', 'Animation', "Children's", 'Comedy', 'Crime', 'Documentary', 'Drama', 
                    'Fantasy', 'Film-Noir', 'Horror', 'Musical', 'Mystery', 'Romance', 'Sci-Fi', 'Thriller', 'War', 'Western'])
invalid_genres = movies['Genres'].str.split('|').explode().unique()
invalid_genres = set(invalid_genres) - valid_genres
print(invalid_genres)

Movies with invalid Genres:
set()


In [77]:
print("Duplicate entries in movies data:", movies.duplicated().sum())

Duplicate entries in movies data: 0


We can conclude that: 
* There are no NULL values in the movies dataframe
* There are no fully same movie titles in the dataset
* The most popular genre is 'Drama'
* All Genre values are valid
* There are no duplicates in the table 

### Let's 

In [ ]:

# Visualizations

# Rating distribution
plt.figure(figsize=(8, 6))
sns.countplot(x='Rating', data=ratings, palette='viridis')
plt.title('Rating Distribution')
plt.xlabel('Rating')
plt.ylabel('Count')
plt.show()

# Ratings per Gender
plt.figure(figsize=(8, 6))
sns.countplot(x='Rating', hue='Gender', data=data, palette='viridis')
plt.title('Rating Distribution by Gender')
plt.xlabel('Rating')
plt.ylabel('Count')
plt.show()

# Ratings per Age Group
plt.figure(figsize=(12, 6))
sns.countplot(x='Rating', hue='Age', data=data, palette='viridis')
plt.title('Rating Distribution by Age Group')
plt.xlabel('Rating')
plt.ylabel('Count')
plt.show()

# Number of ratings per movie
ratings_per_movie = ratings.groupby('MovieID').size()
plt.figure(figsize=(8, 6))
sns.histplot(ratings_per_movie, bins=50, kde=False, color='skyblue')
plt.title('Number of Ratings per Movie')
plt.xlabel('Number of Ratings')
plt.ylabel('Number of Movies')
plt.show()

# Number of ratings per user
ratings_per_user = ratings.groupby('UserID').size()
plt.figure(figsize=(8, 6))
sns.histplot(ratings_per_user, bins=50, kde=False, color='lightgreen')
plt.title('Number of Ratings per User')
plt.xlabel('Number of Ratings')
plt.ylabel('Number of Users')
plt.show()

# Distribution of movie genres
genre_count = movies['Genres'].str.split('|').explode().value_counts()
plt.figure(figsize=(12, 6))
sns.barplot(x=genre_count.index, y=genre_count.values, palette='viridis')
plt.title('Distribution of Movie Genres')
plt.xlabel('Genre')
plt.ylabel('Count')
plt.xticks(rotation=45)
plt.show()
